### PR TITLE
Add python diagrams from GitHub

### LINKED ISSUE #101532: Packaging request: python diagrams library
**Project description**
Generate architecture diagrams with python code. Allows for rapid iteration.

**Metadata**

* homepage URL: https://diagrams.mingrammer.com/
* source URL: https://github.com/mingrammer/diagrams
* license: mit
* platforms: linux, darwin

This library requires graphviz (both native and the python library).

### PR DESCRIPTION
**Pull request by addict3d**
###### Motivation for this change

_See original PR: https://github.com/NixOS/nixpkgs/pull/101533 -- Packaging from GH instead of PyPI to get tests running -- plus its more direct._

Add python diagrams library at version `0.19.1`. Generate nice architecture diagrams
with python code, allowing for rapid iteration.

The images are rendered using dot from graphviz. It outputs your choice of [png jpg svg pdf]

Closes #101532

Thank you @mweinelt 🚀 for all of the help getting this packaged.

###### Things done

I generated some diagrams using this library, with both of the following shell expressions. There are many examples on the project's website.

nix-shell --pure -I nixpkgs=./ -p python3 python3Packages.diagrams python3Packages.graphviz
nix-shell -I nixpkgs=./ -E 'with import <nixpkgs> {}; (python3.withPackages (ps: [ps.diagrams ps.graphviz])).env'

The closure size is ~218M.
The library itself has 22M in images.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/diagrams/default.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  # the deps specified here below dont exactly match what is in pyproject.toml
+  # I did not investigate whether/why the others are listed yet:
+  #
+  #[tool.poetry.dev-dependencies]
+  #pylint = "^2.4"
+  #rope = "^0.14.0"
+  #isort = "^4.3"
+
+  checkInputs = [ pytest ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/diagrams/default.nix:47]: That seems wrong, when the checkPhase is actually using `python -m unittest`.

File: pkgs/development/python-modules/diagrams/default.nix [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  nativeBuildInputs = [ black jinja2 inkscape imagemagick round ];
```
Comment (mweinelt) [inline pkgs/development/python-modules/diagrams/default.nix:48]: The `nativeBuildInputs` are only available during the build. That does not seem sensible for most of these.

- black is an optioniated reformatter for python code, we don't need that in packaging at all
- jinja2 is a templating library, needs to be `propagatedBuildInputs`
- inkscape, imagemagick and round … not sure, likely `buildInputs` or `propagatedBuildInputs` when that doesn't work

**Review by mweinelt — COMMENTED** [review]
Yeah, that is the general direction we should be heading.

**Comment by SuperSandro2000** [comment]
@addict3d friendly ping

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 102060` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>diagrams (python38Packages.diagrams)</li>
    <li>python37Packages.diagrams</li>
    <li>round</li>
  </ul>
</details>

File: pkgs/applications/graphics/round/default.nix [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  version = "0.0.2";
+
+  src = fetchFromGitHub {
+    owner = "mingrammer";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "09brjr3h4qnhlidxlki1by5anahxy16ai078zm4k7ryl579amzdw";
+  };
+
+  #vendorSha256 = "17v5nzswaaaabawxrvkza8a5d9aqyhb2gkncp3fqv6rszmkwwww6";  # there isn't any?
```
Comment (andir) [inline pkgs/applications/graphics/round/default.nix:17]: Is that a question or a statement (both here, below this line and in the other file)?

If there aren't any dependencies why aren't you removing the commented out line with the vendorSha256?

File: pkgs/development/python-modules/diagrams/default.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    owner = "mingrammer";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "1h5i9xb4fsmy81sznvqycxl712d9k695vznkpx2wm1ki7hvz86s8";
+  };
+
+  preConfigure = ''
+    sed -i "s@#!/bin/bash@$!/${bash}/bin/bash@" autogen.sh
+    ./autogen.sh
+  '';
```
Comment (jonringer) [inline pkgs/development/python-modules/diagrams/default.nix:32]:
```suggestion
  preConfigure = ''
    patchShebang autogen.sh
    ./autogen.sh
  '';
```
this should alsow work

File: pkgs/development/python-modules/diagrams/default.nix (hunk 1 above)
Comment (jonringer) [reply pkgs/development/python-modules/diagrams/default.nix:47]: pytest is compatible, but should be
```suggestion
  checkInputs = [ pytestCheckHook ];
```
to ensure tests are running.

However, the current checkphase doesn't use it all... so ?

File: pkgs/applications/graphics/round/default.nix (hunk 3 above)
Comment (addict3d) [reply pkgs/applications/graphics/round/default.nix:17]: It was a statement. Sorry, I PR'ed this is a messy state. Cleaning

File: pkgs/development/python-modules/diagrams/default.nix (hunk 1 above)
Comment (addict3d) [reply pkgs/development/python-modules/diagrams/default.nix:47]: Thanks for the suggestions. Tests are running now (about to push a rebased commit).

`pytest` runs the `unittest`-based tests, after adding `pytestCheckHook` and removing the `checkPhase` (which I had duplicated from the diagrams instructions).

File: pkgs/development/python-modules/diagrams/default.nix (hunk 2 above)
Comment (addict3d) [reply pkgs/development/python-modules/diagrams/default.nix]: From what I can tell during testing and by digging into the project structure, the only runtime dependency is `graphviz`. 

I am testing with `nix-shell --pure -I nixpkgs=./ -p python3 python3Packages.diagrams python3Packages.graphviz` (from my local nixpkgs clone).

The build procedure goes through the list of providers, some get special processing like AWS:
1. [autogen.sh preprocessing AWS provider resources](https://github.com/mingrammer/diagrams/blob/master/autogen.sh#L45)
2. [calls resource script](https://github.com/mingrammer/diagrams/blob/master/scripts/resource.py#L157)
3. [calls round program](https://github.com/mingrammer/diagrams/blob/master/config.py#L22)

Then the [generate script](https://github.com/mingrammer/diagrams/blob/master/scripts/generate.py#L5) is [run for all providers](https://github.com/mingrammer/diagrams/blob/master/autogen.sh#L52).

* black could possibly be removed. It runs last after code generation; [see ref](https://github.com/mingrammer/diagrams/blob/master/autogen.sh#L61). My arguments for keeping are that I struggled to patch it out, and also it looks nicer if someone browses the code.
* jinja2 is build time, used to generate code. There are not any uses at runtime that I saw. I grepped 'jinja' and [saw one instance in python code](https://github.com/mingrammer/diagrams/blob/master/scripts/generate.py#L5).

**Comment by addict3d** [comment]
@martinbaillie Would you mind trying this branch out?

It's the latest release, 0.19.1, and builds directly from the [upstream sources](https://github.com/mingrammer/diagrams). This way we get tests and it will be easier to maintain.

File: pkgs/development/python-modules/diagrams/default.nix [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+  patches = [ ./build_poetry.patch ];
+
+  checkInputs = [ pytestCheckHook ];
+
+  # Despite living in 'tool.poetry.dependencies',
+  # these are only used at build time to process the image resource files
+  nativeBuildInputs = [ black inkscape imagemagick jinja2 poetry round ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/diagrams/default.nix:39]: Can we use poetry-core?

**Review by SuperSandro2000 — COMMENTED** [review]
Please put each package into its own init commit.

File: pkgs/development/python-modules/diagrams/default.nix (hunk 5 above)
Comment (addict3d) [reply pkgs/development/python-modules/diagrams/default.nix]: yes

**Comment by addict3d** [comment]
> Please put each package into its own init commit.

Done. I pushed a new rebase to fix the commit message for the `round` program.

**Comment by SuperSandro2000** [comment]
It is not generating any executable under diagrams bin directory.

**Comment by addict3d** [comment]
> It is not generating any executable under diagrams bin directory.

The library works. I removed the top-level application. Is that fair?

The app can always be added back in later. I don't time to dig in to that right now.

**Pull request merged**